[Feature/v1 007 fetch] Added fetch logical for token, userid, data and render.

NikitaKhadnevich opened the pull request:
This pull request include: 

1. Added fetch token & userId logical;
2. Create some new feature on backend (with @SaXaPhonist );
3. Implement cookie request;
4. Fix wrongs;
5. Rewrote exist method/functions
6. Added fetch userData for render in Profile
7. Profile is working)
We didnt create state manager for save `userid` thus and so `userid` temporarily  saved in cookies

k-krasilnikova commented on frontend/src/api/authConstants.tsx:
```diff
@@ -0,0 +1,23 @@
+};
+
+const TOKEN_FETCH_URL = 'http://localhost:3001/api/account/login';
+const REACT_APP_BACKEND_URL = 'http://localhost:3001';
```
Set it please inside `env` file and reuse env variable

NikitaKhadnevich replied:
Fixed

k-krasilnikova commented on frontend/src/api/authConstants.tsx:
```diff
@@ -0,0 +1,23 @@
+  uniqUserId: 'userId',
+};
+
+const TOKEN_FETCH_URL = 'http://localhost:3001/api/account/login';
```
Is a joke? We already have constants with API routes

NikitaKhadnevich replied:
Fixed

k-krasilnikova commented on frontend/src/api/authConstants.tsx:
```diff
@@ -0,0 +1,23 @@
+  postError: 'Smth went wrong - 500',
+};
+
+const COOKIESVALUES: cookiesTypes = {
```
COOKIES_VALUES

NikitaKhadnevich replied:
Fixed

k-krasilnikova commented on frontend/src/api/authConstants.tsx:
```diff
@@ -0,0 +1,23 @@
+  readonly [key: string]: string;
+}
+
+const REQUESTERRORS: requestErrors = {
```
REQUEST_ERRORS

NikitaKhadnevich replied:
Fixed

k-krasilnikova commented on frontend/src/api/authConstants.tsx:
```diff
@@ -0,0 +1,23 @@
+const TOKEN_FETCH_URL = 'http://localhost:3001/api/account/login';
+const REACT_APP_BACKEND_URL = 'http://localhost:3001';
+
+const { getError, postError } = REQUESTERRORS;
+const { uniqAccessToken, uniqUserId } = COOKIESVALUES;
+export { getError, postError, uniqAccessToken, uniqUserId, TOKEN_FETCH_URL, REACT_APP_BACKEND_URL };
```
Again huge amount of unexpected descructions and exports

NikitaKhadnevich replied:
Remove useless exports

k-krasilnikova commented on frontend/src/api/base.ts:
```diff
@@ -2,14 +2,16 @@ import axios from 'axios';
 import { QueryClient } from 'react-query';
 
 import accessToken from 'constants/accessToken';
+import { REACT_APP_BACKEND_URL } from './authConstants';
```
Filter please imports

NikitaKhadnevich replied:
Rewrote paths

k-krasilnikova commented on frontend/src/api/base.ts:
```diff
@@ -2,14 +2,16 @@ import axios from 'axios';
   ? axios.create({
-      baseURL: process.env.REACT_APP_BACKEND_URL,
+      // baseURL: process.env.REACT_APP_BACKEND_URL, Wrong Path - 3000
+      baseURL: REACT_APP_BACKEND_URL,
```
env variable!

NikitaKhadnevich replied:
Fixed

k-krasilnikova commented on frontend/src/api/profile/getToken.tsx:
```diff
@@ -0,0 +1,7 @@
+import { useMutation } from 'react-query';
+import putUserToken from '../restApi';
```
Filter imports

k-krasilnikova replied:
Remove please this file. It's an auth module, not profile!

NikitaKhadnevich replied:
Removed and created correct folder/file

k-krasilnikova commented on frontend/src/api/restApi.tsx:
```diff
@@ -0,0 +1,22 @@
+import axios from 'axios';
+import Cookies from 'js-cookie';
+
+import { postError, uniqAccessToken, uniqUserId } from './authConstants';
+
+const putUserToken = async (
+  baseUrl: string,
+  initialData: string | unknown,
+): Promise<Response | Error | void> => {
+  try {
+    const response = await axios.post(baseUrl, initialData);
+    const responseData = await response.data;
+    const resAccess = JSON.stringify(responseData.accessToken);
+    const resUserId = JSON.stringify(responseData._id);
+    Cookies.set(uniqAccessToken, resAccess, { secure: true });
+    Cookies.set(uniqUserId, resUserId, { secure: true });
+  } catch (error) {
+    return new Error(postError);
+  }
+};
+
+export default putUserToken;
```
Remove this file and rewrite please this logic in style of getting client profile!

NikitaKhadnevich replied:
Rewrote this

k-krasilnikova commented on frontend/src/constants/accessToken.ts:
```diff
@@ -1,5 +1,8 @@
-const accessToken = Cookies.get('accessToken');
+const token = Cookies.get(uniqAccessToken);
+let accessToken = '';
+token ? (accessToken = JSON.parse(token)) : (accessToken = '');
```
Wrong definition of logic with assigning variable value. Rewrite it please with something more clear and readable.

NikitaKhadnevich replied:
Created universal function for get cookies

k-krasilnikova commented on frontend/src/constants/routes.ts:
```diff
@@ -23,5 +25,5 @@ export const PATHS: {
 export const API: {
   getProfile: string;
 } = {
-  getProfile: '/api/account/profile',
+  getProfile: `/api/users/${userId}`,
```
Are you sure that we should replace this route?

NikitaKhadnevich replied:
I suppose, In present we will work with this

k-krasilnikova commented on frontend/src/pages/CoursesList/CoursesList.tsx:
```diff
@@ -4,7 +4,7 @@ import { AuthorizedLayout } from 'components/Layout';
 const CoursesList: React.FC = () => (
   <AuthorizedLayout pageName="Courses List">
-    <div>Courses List should be here</div>
+    <div>Should be Courses List</div>
```
Any reasons for such replacing?

NikitaKhadnevich replied:
Fixed

k-krasilnikova commented on frontend/src/pages/Profile/ProfileContainer.tsx:
```diff
@@ -1,11 +1,11 @@
 
 const ProfileContainer: React.FC = () => {
-  const { data, isSuccess, isLoading, isError } = useGetProfile();
+  const { data } = useGetProfile();
```
Do we really need to remove those props?

NikitaKhadnevich replied:
If we will safe it, ts will show warning 'yellow' message as like ` isSuccess, isLoading, isError dnt use`

k-krasilnikova commented on frontend/src/pages/SignIn/SignInContainer.tsx:
```diff
@@ -16,11 +18,16 @@ const initSignInvalue: SignInFields = {
 };
 
 const SignInContainer: React.FC = () => {
+  const navToAuth = useNavigate();
```
What does this name mean?

NikitaKhadnevich replied:
Renamed

NikitaKhadnevich commented:
This update contain next changes: 

1. Fixed file names, folders and removed/delete some of them;
2. Added new function for work with cookies;
3. implement this in `pages` (sign in profile) and also `PrivateRoute` component
4. rewrote custom `mutate` hook as like custom `useQuery` in our api folder

k-krasilnikova commented on frontend/src/api/auth/getAuth.tsx:
```diff
@@ -0,0 +1,30 @@
+
+    try {
+      const response = await apiClient.post(API.getToken, initialData);
+      tokenResponse = await response.data;
```
Do we really need double `await` here?

NikitaKhadnevich replied:
Rewrote this

k-krasilnikova commented on frontend/src/api/base.ts:
```diff
@@ -1,17 +1,21 @@
 import { QueryClient } from 'react-query';
 
-import accessToken from 'constants/accessToken';
+import { getAuthResponseData } from 'utils/Helpers/getAuthResponseData';
```
Why `helpers` from uppercase?

NikitaKhadnevich replied:
renamed

k-krasilnikova commented on frontend/src/utils/Helpers/getAuthResponseData.tsx:
```diff
@@ -0,0 +1,6 @@
+
+export const getAuthResponseData = (cookieName: string): string | undefined => {
+  const backendData: string | undefined = Cookies.get(cookieName);
+  if (backendData) return JSON.parse(backendData);
```
Why helpers from uppercase?
Keep please clear structure of `if` statement!

NikitaKhadnevich replied:
fixed

SaXaPhonist commented on frontend/src/api/auth/getAuth.tsx:
```diff
@@ -0,0 +1,30 @@
+
+const useGetAuth = () =>
+  useMutation(async (initialData: string | unknown) => {
+    let tokenResponse: {
```
you can create tokenResopnse interface

NikitaKhadnevich replied:
We have tokenResponse inside `useMutation`. Let's save this?

SaXaPhonist commented on frontend/src/pages/SignIn/SignInContainer.tsx:
```diff
@@ -16,11 +17,16 @@ const initSignInvalue: SignInFields = {
     initialValues: initSignInvalue,
     validationSchema: signInSchema,
     onSubmit: (values, { resetForm }): void => {
+      auth.mutateAsync(values);
```
await

NikitaKhadnevich replied:
`mutateAsync`  already `async`

SaXaPhonist replied:
that is the point

NikitaKhadnevich commented on backend/src/modules/auth/authController.ts:
```diff
@@ -45,6 +47,7 @@ const refreshController = async (req: Request, res: Response, next: TMiddlewareC
     }
 
     const newTokens = await generateJWT(dbUser);
+    res.cookie('refreshToken', newTokens.refreshToken, {maxAge: TIME_30D_SEC});
```
Added this

NikitaKhadnevich commented on frontend/src/api/auth/getAuth.tsx:
```diff
@@ -0,0 +1,30 @@
+      Cookies.set(COOKIE_VALUES.uniqAccessToken, resAccess, { secure: true });
+      Cookies.set(COOKIE_VALUES.uniqUserId, resUserId, { secure: true });
+    } catch (error) {
+      tokenResponse = { error: error };
```
Added new Error

k-krasilnikova commented on frontend/src/api/auth/getAuth.tsx:
```diff
@@ -0,0 +1,30 @@
+    try {
+      const response = await apiClient.post(API.getToken, initialData);
+      tokenResponse = response.data;
+      const resAccess = JSON.stringify(tokenResponse.accessToken);
+      const resUserId = JSON.stringify(tokenResponse._id);
```
Rename please contants to something more clear, example: `accessToken` and `userId`

NikitaKhadnevich replied:
Renamed

k-krasilnikova commented on frontend/src/api/profile/getProfile.ts:
```diff
@@ -16,13 +17,14 @@ const useGetProfile = () =>
       skype?: string;
       error?: unknown;
     };
+    const apiClient = apiClientWrapper(COOKIE_VALUES?.uniqAccessToken);
```
Let's use getting of the token directly

NikitaKhadnevich replied:
Cleared this

k-krasilnikova commented on frontend/src/components/PrivateRoute/PrivateRoute.tsx:
```diff
@@ -2,9 +2,11 @@ import React from 'react';
+import { COOKIE_VALUES } from 'constants/authConstants';
 
 const PrivateRoute = () => {
+  const accessToken = getAuthResponseData(COOKIE_VALUES?.uniqAccessToken);
```
!

NikitaKhadnevich replied:
Also cleared. Added new helper-selector function

k-krasilnikova commented on frontend/src/pages/SignIn/SignInContainer.tsx:
```diff
@@ -16,11 +17,16 @@ const initSignInvalue: SignInFields = {
 
 const SignInContainer: React.FC = () => {
+  const goToProfile = useNavigate();
+  const auth = useGetAuth();
```
Destructure please auth the same way as it was in example that I've send you
```suggestion
  const { mutateAsync } = useGetAuth();
```

NikitaKhadnevich replied:
Fixed

k-krasilnikova commented on frontend/src/pages/SignIn/SignInContainer.tsx:
```diff
@@ -16,11 +17,16 @@ const initSignInvalue: SignInFields = {
     onSubmit: (values, { resetForm }): void => {
+      auth.mutateAsync(values);
       resetForm();
+      goToProfile(PATHS.profile);
```
It's the wrong name for the function. In case it's going to the profile, why do we need to send the same route additionally
```suggestion
      useNavigate(PATHS.profile);
```

k-krasilnikova replied:
Also we should redirect client only in case of success! Do we need to remove this call to the onSuccess?

NikitaKhadnevich replied:
Wep, now `onSuccess + redirect`  implemented and working. Renamed const `goToProfile` to `navigateTo` , useNavigate cannt receive args

NikitaKhadnevich commented:
Fixed conflicts, recreated `helpers` folders, added some backend logical.
